### DISCUSSION TITLE
Typescript with Modals

### DISCUSSION THREAD

**Discussion by 3baaady07** [twbs/bootstrap#39538] [Q&A] [answered]
I'm trying to create a [Varying modal content](https://getbootstrap.com/docs/5.3/components/modal/#varying-modal-content) using typescript. But the following code keeps failing with the error
```
Argument of type '"show.bs.modal"' is not assignable to parameter of type 'keyof HTMLElementEventMap'.
```

```javascript
// javascript
const exampleModal = document.getElementById('exampleModal')
if (exampleModal) {
  exampleModal.addEventListener('show.bs.modal', event => {
    // Button that triggered the modal
    const button = event.relatedTarget
    // Extract info from data-bs-* attributes
    const recipient = button.getAttribute('data-bs-whatever')
    // If necessary, you could initiate an Ajax request here
    // and then do the updating in a callback.

    // Update the modal's content.
    const modalTitle = exampleModal.querySelector('.modal-title')
    const modalBodyInput = exampleModal.querySelector('.modal-body input')

    modalTitle.textContent = `New message to ${recipient}`
    modalBodyInput.value = recipient
  })
}
```

I've been looking for a fix to this for quite sometime now...

Cheers,

**Comment by julien-deramond** [comment, marked as answer]
Hey @3baaady07, could you please create a reproducible test case using StackBlitz or any other relevant tools? Ensuring we're on the same page with the configuration and tools will greatly assist in finding a solution together.

**Reply by 3baaady07** [reply]
Funny that your request urged me to to isolate the issue and, as a result, find a solution 😁

Here's the [StackBlitz](https://stackblitz.com/github/3baaady07/modal-typescript) solution.

Initially, I had the following code to handle varying modal content.
```ts
interface HTMLElementEventMap
{
   ["show.bs.modal"]: MouseEvent
}

modal.addEventListener("show.bs.modal", (e: HTMLElementEventMap["show.bs.modal"]) => {
   const invoker = e.relatedTarget as HTMLButtonElement // <-- didn't work
   // rest of the code ...
})
```
It turns out, I didn't need to enforce a type on the `e` argument as it is, by default, an `Event` type. Then within the handler block, `e` should be "type-casted" _(is that a word?)_ to `MouseEvent`.

The fix is...
```ts
modal.addEventListener("show.bs.modal", (e) => {
   const event = e as MouseEvent;
   const invoker = event.relatedTarget as HTMLButtonElement
   // rest of the code ...
})
```

Anyways, it's solved now.

Thanks @julien-deramond 👍

**Reply by julien-deramond** [reply]
Thanks @3baaady07!
Appreciate your dedication in providing a StackBlitz reproducible environment along with a thorough analysis. This contribution will prove invaluable for folks encountering similar issues in the future!

**Reply by 3baaady07** [reply]
My pleasure. I appreciate the work you do on this project. I'll gladly help in anyway I can.

Cheers,
